### ISSUE TITLE
read -n bug

### ISSUE THREAD

**Issue by andychu** [oilshell/oil#438]
cc @jyn514 

I was trying to come up with a test case for your patch... and I found this problem.  It's probably related, but I haven't looked deeply into it.

I think the patch is probably good, but having a test case would be nice.

```
$ bash -c 'while true; do echo -n x; done | { read -n 3; echo $REPLY; }'    
xxx
$ osh -c 'while true; do echo -n x; done | { read -n 3; echo $REPLY; }'     
x
```

**Comment by jyn514** [comment]
The issue with automated testing is that all the (non-interactive) ways I can think of to get input to `osh` will also send EOF, which defeats the point of the test. I guess I could mess around with named pipes to avoid that? Not sure how portable those are.

**Comment by andychu** [comment]
That's why I have an infinite loop there.  No EOF :)

The right hand process dies, which I think causes the the left hand process to be killed with SIGPIPE.

**Comment by jyn514** [comment]
Hmm, I think this is because `posix.read` can return a different number of bytes than we asked for. We should put that in a loop until we have the wanted number of bytes.

This is only semi-related to my patch, non-interactive input shouldn't hit the new code path at all.

**Comment by jyn514** [comment]
Yup I have a fix for that too :)

**Comment by jyn514** [comment]
Hmm all other shells exit with `124`, not sure why osh doesn't.

**Comment by jyn514** [comment]
This should be fixed since you merged https://github.com/oilshell/oil/pull/437.

**Comment by andychu** [comment]
Released with 0.7.pre2: https://www.oilshell.org/release/0.7.pre2/
